### PR TITLE
replace bespoke expression evaluator with one in fortran-src; use more precise array dimension encoding

### PR DESCRIPTION
**Pull request by raehik**
fortran-vars contains a restricted Fortran expression evaluator. This is used for turning constant expressions into their compile-time value, and gathering extra information by attempting to evaluate expressions in certain contexts to see if they are constant. Rather than using `Language.Fortran.AST.Value` (which includes syntactic features irrelevant here), compile-time values are stored in a custom value representation type.

fortran-src now has its own value representation and expression evaluator (inspired by the one in fortran-vars). It stores more precise information about values and emits lots of debug info.

This PR swaps out fortran-vars' expression evaluator for fortran-src. Note that values are still stored using the fortran-vars representation -- when the evaluator is called, values are converted on the fly to the new fortran-src representation, and results are converted back. This goes for the unit tests too.

We also update how array variable dimensions are encoded. Code that touches arrays is updated. There may be generalizations and improvements to analyzing more code -- I didn't focus on finding those. See https://github.com/camfort/fortran-src/pull/261 and https://github.com/camfort/fortran-src/issues/254 for further details.

Currently, there remains lots of unused code and things to clean up (delete, rename). In particular, the original evaluator still exists.

To-do:

  * [x] clean up history: shift Nix additions (merge and move to end), merge small commits

### CODE REVIEW

File: src/Language/Fortran/Vars/Eval.hs [hunk 1]
```diff
@@ -30,22 +34,37 @@ import           Language.Fortran.Vars.Types    ( SymbolTableEntry(..)
...
+eval' :: SymbolTable -> Expression a -> Either String ExpVal
+eval' = eval'ViaFS
+
 -- | Given a 'SymbolTable' and some 'Expression', evaluate that expression
 -- into a basic type and return it as an 'ExpVal' or a 'String' describing
 -- the issue that prevented the evaluation
-eval' :: SymbolTable -> Expression a -> Either String ExpVal
-eval' symTable expr = case expr of
-  ExpValue _ _ (ValVariable name) -> case M.lookup name symTable of
+eval'Orig :: SymbolTable -> Expression a -> Either String ExpVal
```
Comment (dorchard) [inline src/Language/Fortran/Vars/Eval.hs]: Do we need to keep this? Or is it here for the purposes of testing between the two implementations (doesn't look like there are any such tests though).

File: src/Language/Fortran/Vars/FSRepr/Eval.hs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+-- | Replacement module for 'Language.Fortran.Vars.Eval' that wraps the existing
+--   interface to use the strong representation internally.
+module Language.Fortran.Vars.FSRepr.Eval where
+
+{-
```
Comment (dorchard) [inline src/Language/Fortran/Vars/FSRepr/Eval.hs]: Can we remove these comments?

File: src/Language/Fortran/Vars/Eval.hs (hunk 1 above)
Comment (raehik) [reply src/Language/Fortran/Vars/Eval.hs]: That should go, I wrote it like that to stay non-invasive and swap between evaluators.

File: src/Language/Fortran/Vars/FSRepr/Eval.hs (hunk 2 above)
Comment (raehik) [reply src/Language/Fortran/Vars/FSRepr/Eval.hs]: Done.

**Comment by RaoulHC** [comment]
This all looks sensible to me.

Just to check the unhandled boz coercions will just mean that those symbols are left in the symbol table as `SVariable`s right? Which I think is fine.

**Comment by raehik** [comment]
> Just to check the unhandled boz coercions will just mean that those symbols are left in the symbol table as `SVariable`s right? Which I think is fine.

Yep. We only handle INTEGERs, so if you try to assign a BOZ directly to a a REAL, it won't do anything. A quick check shows [previous code](https://github.com/camfort/fortran-vars/blob/e6db60bcd48ccf49f3de1cde584117f6a0c5f03f/src/Language/Fortran/Vars/BozConstant.hs#L66-L73) would runtime error instead, so similar fail cases.

This special case will only occur for declarations of form `INTEGER a; PARAMETER(a =  b'01011101')`. Still compiles without, but some BOZ tests asserting `INT*2`, `INT*4` bounds break.

**Pull request merged**